### ISSUE TITLE
support Azure OpenAI api

### ISSUE THREAD

**Issue by wxxb789** [karthink/gptel#41]
will this package support Azure OpenAI api?

**Comment by karthink** [comment]
Not explicitly, no. But I had a look at the API:

``` shell
curl https://YOUR_RESOURCE_NAME.openai.azure.com/openai/deployments/YOUR_DEPLOYMENT_NAME/chat/completions?api-version=2023-03-15-preview \
  -H "Content-Type: application/json" \
  -H "api-key: YOUR_API_KEY" \
  -d '{"messages":[{"role": "system", "content": "You are a helpful assistant."},{"role": "user", "content": "Does Azure OpenAI support customer managed keys?"},{"role": "assistant", "content": "Yes, customer managed keys are supported by Azure OpenAI."},{"role": "user", "content": "Do other Azure Cognitive Services support this too?"}]}'
```

It looks very similar to the OpenAI API, so you might be able to use gptel with minimal tweaking: [Replace the request URL](https://github.com/karthink/gptel/blob/master/gptel-curl.el#L45) and one of the headers (`Authorization: Bearer sk-...` → `api-key: sk-...`). If that works I could add a PR.

**Comment by jcs090218** [comment]
Maybe worth mentioning, [openai.el](https://github.com/emacs-openai/openai/ now supports this. 😄 See https://github.com/emacs-openai/openai/issues/14.

**Comment by karthink** [comment]
> Maybe worth mentioning, [openai.el](https://github.com/emacs-openai/openai/) now supports this. smile See [emacs-openai/openai#14](https://github.com/emacs-openai/openai/issues/14).

Thanks for the heads-up. If all that's needed is making the base url configurable, this should be simple to add.

**Comment by TeMPOraL** [comment]
It's not _that_ simple - see https://github.com/emacs-openai/openai/issues/18.
Basically, Azure OpenAI API is slightly different:

`POST https://YOUR_RESOURCE_NAME.openai.azure.com/openai/deployments/YOUR_DEPLOYMENT_NAME/completions?api-version=2022-12-01`

In particular, you:
- Don't pass `model` parameter in the payload, but istead it's the `YOUR_DEPLOYMENT_NAME`;
- You don't have `v1`, but a `api-version` URL parameter;
- You don't pass the API key as `Authorization: Bearer $key`, but as `api-key: $key` header.

I would dearly love to see this implemented too.

**Comment by karthink** [comment]
Support for Azure has been implemented in v0.5.0, along with support for local models using GPT4All and Ollama.
